src/libfsm/dfavm.c

```diff
@@ -0,0 +1,1692 @@
+
+		(void)running_print_op;  /* make clang happy */
+#if DEBUG_VM_EXECUTION
+		running_print_op(vm->ops, st->pc, sp, buf, n, ch, stderr);
+#endif /* DEBUG_VM_EXECUTION */
+
+		if (op == VM_OP_FETCH) {
+			int end;
+
+			end = b & 0x01;
+			if (sp >= last) {
+				st->fetch_state = end;
+				return VM_MATCHING;
+			}
+
+			ch = (unsigned char) *sp++;
+			st->pc++;
+		} else {
+			int cmp, end, dest, dest_nbytes;
+			int result;
+			uint32_t off;
+
+			/* either STOP or BRANCH */
+			off = st->pc+1;
+			cmp  = b >> 5;
+
+			result = 0;
+			if (cmp == VM_CMP_ALWAYS) {
+				result = 1;
+			} else {
+				int arg;
+
+				arg = vm->ops[off++];
+
+				switch (cmp) {
+				case VM_CMP_LT: result = ch <  arg; break;
+				case VM_CMP_LE: result = ch <= arg; break;
+				case VM_CMP_GE: result = ch >= arg; break;
+				case VM_CMP_GT: result = ch >  arg; break;
+				case VM_CMP_EQ: result = ch == arg; break;
+				case VM_CMP_NE: result = ch != arg; break;
+				}
+			}
+
+			dest = b & 0x3;
+			dest_nbytes = 1 << dest;
+
+			if (result) {
+				int32_t rel;
+
+				if (op == VM_OP_STOP) {
+					end = b & 0x01;
+					st->state = (end == VM_END_FAIL) ? VM_FAIL : VM_SUCCESS;
+					return st->state;
+				}
+
+				if (dest == 0) {
+					rel = (int8_t)(vm->ops[off++]);
+				} else if (dest == 1) {
+					union {
+						uint16_t u;
+						int16_t  i;
+					} packed;
+
+					packed.u = (vm->ops[off+0] | (vm->ops[off+1] << 8));
+					rel = packed.i;
+					off += 2;
+				} else {
+					union {
+						uint32_t u;
+						int32_t  i;
+					} packed;
+					packed.u = vm->ops[off+0] | (vm->ops[off+1] << 8) | (vm->ops[off+2] << 16) | (vm->ops[off+3] << 24);
+					rel = packed.i;
+				}
+
+				st->pc += rel;
+			} else if (op == VM_OP_BRANCH) {
+				st->pc = off + dest_nbytes;
+			} else {
+				st->pc = off;
+			}
+		}
+	}
+
+	return VM_FAIL;
+}
+
+static enum dfavm_state 
+vm_end(const struct fsm_dfavm *vm, struct vm_state *st)
+{
+	(void)vm;
+
+#if DEBUG_VM_EXECUTION
+	fprintf(stderr, "END: st->state = %d, st->fetch_state = %d\n",
+		st->state, st->fetch_state);
+#endif /* DEBUG_VM_EXECUTION */
+
+	if (st->state != VM_MATCHING) {
+		return st->state;
+	}
+
+	return (st->fetch_state == VM_END_SUCC) ? VM_SUCCESS : VM_FAIL;
+}
+
+static const struct vm_state state_init;
+
+int
+fsm_vm_match_file(const struct fsm_dfavm *vm, FILE *f)
+{
+	struct vm_state st = state_init;
+	enum dfavm_state result;
+	char buf[4096];
+
+	result = VM_FAIL;
+	for (;;) {
+		size_t nb;
+
+		nb = fread(buf, 1, sizeof buf, f);
+		if (nb == 0) {
+			break;
+		}
+
+		result = vm_match(vm, &st, buf, nb);
+		if (result != VM_MATCHING) {
+			return result == VM_SUCCESS;
+		}
+	}
+
+	if (ferror(f)) {
+		return 0;
+	}
+
+	return vm_end(vm, &st) == VM_SUCCESS;
+}
+
+int
+fsm_vm_match_buffer(const struct fsm_dfavm *vm, const char *buf, size_t n)
+{
+	struct vm_state st = state_init;
+	enum dfavm_state result;
+
+	vm_match(vm, &st, buf, n);
+	result = vm_end(vm, &st);
+
+	return result == VM_SUCCESS;
+}
```

Comment:
Maybe not on this initial PR, but it'd probably be sensible to split this into several files, for compilation, execution, etc.

Reply:
Yes!  This file is getting too large.